RingbufferAddAllReadManyStressTest.whenShortTTLAndBigBuffer

- Fails on `Hazelcast-3.x-IbmJDK8-fips-nightly`
- Fails on [this build](http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.x-IbmJDK8-fips-nightly/116/testReport/com.hazelcast.ringbuffer.impl/RingbufferAddAllReadManyStressTest/whenShortTTLAndBigBuffer/)
- Error 
```
No error should have been thrown, but ConsumeThread-1 completed error expected null, but was:<java.lang.AssertionError: expected:<183774> but was:<183907>>
```
- Stack Trace
```
java.lang.AssertionError: No error should have been thrown, but ConsumeThread-1 completed error expected null, but was:<java.lang.AssertionError: expected:<183774> but was:<183907>>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotNull(Assert.java:755)
	at org.junit.Assert.assertNull(Assert.java:737)
	at com.hazelcast.test.TestThread.assertSucceedsEventually(TestThread.java:78)
	at com.hazelcast.ringbuffer.impl.RingbufferAddAllReadManyStressTest.test(RingbufferAddAllReadManyStressTest.java:126)
	at com.hazelcast.ringbuffer.impl.RingbufferAddAllReadManyStressTest.whenShortTTLAndBigBuffer(RingbufferAddAllReadManyStressTest.java:100)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:811)
```
- Possible commit: f5328fc

kindly check.

Reopening since it failed again with `4.2.z` branch and IBM JDK 8 - http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-4.maintenance-IbmJDK8-nightly/1/testReport/com.hazelcast.ringbuffer.impl/RingbufferAddAllReadManyStressTest/whenShortTTLAndBigBuffer/

Stacktrace:
```
java.lang.AssertionError: No error should have been thrown, but ConsumeThread-2 completed error expected null, but was:<java.lang.AssertionError: expected:<5586259> but was:<5586465>>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotNull(Assert.java:756)
	at org.junit.Assert.assertNull(Assert.java:738)
	at com.hazelcast.test.TestThread.assertSucceedsEventually(TestThread.java:78)
	at com.hazelcast.ringbuffer.impl.RingbufferAddAllReadManyStressTest.test(RingbufferAddAllReadManyStressTest.java:127)
	at com.hazelcast.ringbuffer.impl.RingbufferAddAllReadManyStressTest.whenShortTTLAndBigBuffer(RingbufferAddAllReadManyStressTest.java:100)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:822)
```

Failed with IBM JDK 8 and fips also on `master` branch: http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-IbmJDK8-fips-nightly/315/testReport/com.hazelcast.ringbuffer.impl/RingbufferAddAllReadManyStressTest/whenShortTTLAndBigBuffer/

Stacktrace:
```
java.lang.AssertionError: No error should have been thrown, but ConsumeThread-1 completed error expected null, but was:<java.lang.AssertionError: expected:<1265089> but was:<1265127>>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotNull(Assert.java:756)
	at org.junit.Assert.assertNull(Assert.java:738)
	at com.hazelcast.test.TestThread.assertSucceedsEventually(TestThread.java:78)
	at com.hazelcast.ringbuffer.impl.RingbufferAddAllReadManyStressTest.test(RingbufferAddAllReadManyStressTest.java:126)
	at com.hazelcast.ringbuffer.impl.RingbufferAddAllReadManyStressTest.whenShortTTLAndBigBuffer(RingbufferAddAllReadManyStressTest.java:100)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:822)
```

Standard output (there's one big pause):
```
Hiccups measured while running test 'whenShortTTLAndBigBuffer(com.hazelcast.ringbuffer.impl.RingbufferAddAllReadManyStressTest):'
06:55:55, accumulated pauses: 66 ms, max pause: 37 ms, pauses over 1000 ms: 0
06:56:00, accumulated pauses: 4699 ms, max pause: 3766 ms, pauses over 1000 ms: 1
06:56:05, accumulated pauses: 5 ms, max pause: 0 ms, pauses over 1000 ms: 0
```

Failed with IBM JDK 8 on `4.2.z` branch: http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-4.maintenance-IbmJDK8-nightly/13/testReport/com.hazelcast.ringbuffer.impl/RingbufferAddAllReadManyStressTest/whenShortTTLAndBigBuffer/

Stacktrace:
```
java.lang.AssertionError: No error should have been thrown, but ConsumeThread-2 completed error expected null, but was:<java.lang.AssertionError: expected:<1256278> but was:<1256283>>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotNull(Assert.java:756)
	at org.junit.Assert.assertNull(Assert.java:738)
	at com.hazelcast.test.TestThread.assertSucceedsEventually(TestThread.java:78)
	at com.hazelcast.ringbuffer.impl.RingbufferAddAllReadManyStressTest.test(RingbufferAddAllReadManyStressTest.java:127)
	at com.hazelcast.ringbuffer.impl.RingbufferAddAllReadManyStressTest.whenShortTTLAndBigBuffer(RingbufferAddAllReadManyStressTest.java:100)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:822)
```

Standard output (there's again one big pause):
```
21:08:18,715  INFO |whenShortTTLAndBigBuffer| - [MetricsConfigHelper] whenShortTTLAndBigBuffer - [LOCAL] [dev] [4.2.3-SNAPSHOT] Overridden metrics configuration with system property 'hazelcast.metrics.collection.frequency'='1' -> 'MetricsConfig.collectionFrequencySeconds'='1'
21:08:18,716  INFO |whenShortTTLAndBigBuffer| - [system] whenShortTTLAndBigBuffer - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] Hazelcast 4.2.3-SNAPSHOT (20210814 - a516020) starting at [127.0.0.1]:5701
21:08:18,717  INFO |whenShortTTLAndBigBuffer| - [MetricsConfigHelper] whenShortTTLAndBigBuffer - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] Collecting debug metrics and sending to diagnostics is enabled
21:08:18,721  WARN |whenShortTTLAndBigBuffer| - [CPSubsystem] whenShortTTLAndBigBuffer - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
21:08:18,731  INFO |whenShortTTLAndBigBuffer| - [Diagnostics] whenShortTTLAndBigBuffer - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
21:08:18,731  INFO |whenShortTTLAndBigBuffer| - [LifecycleService] whenShortTTLAndBigBuffer - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] [127.0.0.1]:5701 is STARTING
21:08:18,731  INFO |whenShortTTLAndBigBuffer| - [ClusterService] whenShortTTLAndBigBuffer - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] 

Members {size:1, ver:1} [
	Member [127.0.0.1]:5701 - 7ed2c071-d33b-45d0-add3-f748f11ab08f this
]

21:08:18,733  INFO |whenShortTTLAndBigBuffer| - [LifecycleService] whenShortTTLAndBigBuffer - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] [127.0.0.1]:5701 is STARTED
21:08:18,733  INFO |whenShortTTLAndBigBuffer| - [MetricsConfigHelper] whenShortTTLAndBigBuffer - [LOCAL] [dev] [4.2.3-SNAPSHOT] Overridden metrics configuration with system property 'hazelcast.metrics.collection.frequency'='1' -> 'MetricsConfig.collectionFrequencySeconds'='1'
21:08:18,733  INFO |whenShortTTLAndBigBuffer| - [system] whenShortTTLAndBigBuffer - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] Hazelcast 4.2.3-SNAPSHOT (20210814 - a516020) starting at [127.0.0.1]:5702
21:08:18,734  INFO |whenShortTTLAndBigBuffer| - [HealthMonitor] hz.gracious_fermi.HealthMonitor - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] processors=8, physical.memory.total=377.6G, physical.memory.free=210.4G, swap.space.total=4.0G, swap.space.free=3.8G, heap.memory.used=3.2G, heap.memory.free=788.0M, heap.memory.total=4.0G, heap.memory.max=4.0G, heap.memory.used/total=80.76%, heap.memory.used/max=80.76%, minor.gc.count=11771, minor.gc.time=113436ms, major.gc.count=2, major.gc.time=3265ms, load.process=1.55%, load.system=15.58%, load.systemAverage=49.46, thread.count=83, thread.peakCount=2061, cluster.timeDiff=0, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.client.query.size=0, executor.q.client.blocking.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operations.size=0, executor.q.priorityOperation.size=0, operations.completed.count=0, executor.q.mapLoad.size=0, executor.q.mapLoadAllKeys.size=0, executor.q.cluster.size=0, executor.q.response.size=0, operations.running.count=0, operations.pending.invocations.percentage=0.00%, operations.pending.invocations.count=0, proxy.count=0, clientEndpoint.count=0, connection.active.count=0, client.connection.count=0, connection.count=0
21:08:18,734  INFO |whenShortTTLAndBigBuffer| - [MetricsConfigHelper] whenShortTTLAndBigBuffer - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] Collecting debug metrics and sending to diagnostics is enabled
21:08:18,737  WARN |whenShortTTLAndBigBuffer| - [CPSubsystem] whenShortTTLAndBigBuffer - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
21:08:18,746  INFO |whenShortTTLAndBigBuffer| - [Diagnostics] whenShortTTLAndBigBuffer - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
21:08:18,746  INFO |whenShortTTLAndBigBuffer| - [LifecycleService] whenShortTTLAndBigBuffer - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] [127.0.0.1]:5702 is STARTING
21:08:18,747  INFO |whenShortTTLAndBigBuffer| - [MockServer] whenShortTTLAndBigBuffer - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] Created connection to endpoint: [127.0.0.1]:5701, connection: MockConnection{localEndpoint=[127.0.0.1]:5702, remoteEndpoint=[127.0.0.1]:5701, alive=true}
21:08:18,748  INFO |whenShortTTLAndBigBuffer| - [MockServer] hz.gracious_fermi.priority-generic-operation.thread-0 - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] Created connection to endpoint: [127.0.0.1]:5702, connection: MockConnection{localEndpoint=[127.0.0.1]:5701, remoteEndpoint=[127.0.0.1]:5702, alive=true}
21:08:18,749  INFO |whenShortTTLAndBigBuffer| - [ClusterService] hz.gracious_fermi.priority-generic-operation.thread-0 - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] 

Members {size:2, ver:2} [
	Member [127.0.0.1]:5701 - 7ed2c071-d33b-45d0-add3-f748f11ab08f this
	Member [127.0.0.1]:5702 - be2ccebd-6e7a-4b16-8363-00899e7e931d
]

21:08:18,847  INFO |whenShortTTLAndBigBuffer| - [ClusterService] hz.wizardly_fermi.priority-generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] 

Members {size:2, ver:2} [
	Member [127.0.0.1]:5701 - 7ed2c071-d33b-45d0-add3-f748f11ab08f
	Member [127.0.0.1]:5702 - be2ccebd-6e7a-4b16-8363-00899e7e931d this
]

21:08:19,250  INFO |whenShortTTLAndBigBuffer| - [LifecycleService] whenShortTTLAndBigBuffer - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] [127.0.0.1]:5702 is STARTED
21:08:19,254  INFO |whenShortTTLAndBigBuffer| - [PartitionStateManager] ConsumeThread-1 - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] Initializing cluster partition table arrangement...
21:08:19,255  INFO |whenShortTTLAndBigBuffer| - [HealthMonitor] hz.wizardly_fermi.HealthMonitor - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] processors=8, physical.memory.total=377.6G, physical.memory.free=210.5G, swap.space.total=4.0G, swap.space.free=3.8G, heap.memory.used=3.2G, heap.memory.free=786.2M, heap.memory.total=4.0G, heap.memory.max=4.0G, heap.memory.used/total=80.81%, heap.memory.used/max=80.81%, minor.gc.count=11771, minor.gc.time=113436ms, major.gc.count=2, major.gc.time=3265ms, load.process=0.08%, load.system=18.42%, load.systemAverage=49.46, thread.count=107, thread.peakCount=2061, cluster.timeDiff=0, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.client.query.size=0, executor.q.client.blocking.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operations.size=0, executor.q.priorityOperation.size=0, operations.completed.count=12, executor.q.mapLoad.size=0, executor.q.mapLoadAllKeys.size=0, executor.q.cluster.size=0, executor.q.response.size=0, operations.running.count=1, operations.pending.invocations.percentage=0.00%, operations.pending.invocations.count=0, proxy.count=1, clientEndpoint.count=0, connection.active.count=0, client.connection.count=0, connection.count=0
21:08:31,510  INFO |whenShortTTLAndBigBuffer| - [LifecycleService] main - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] [127.0.0.1]:5702 is SHUTTING_DOWN
21:08:31,510  WARN |whenShortTTLAndBigBuffer| - [Node] main - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] Terminating forcefully...
21:08:31,510  INFO |whenShortTTLAndBigBuffer| - [Node] main - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] Shutting down connection manager...
21:08:31,510  INFO |whenShortTTLAndBigBuffer| - [MockServer] main - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5702, connection: MockConnection{localEndpoint=[127.0.0.1]:5701, remoteEndpoint=[127.0.0.1]:5702, alive=false}
21:08:31,510  INFO |whenShortTTLAndBigBuffer| - [MockServer] main - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5701, connection: MockConnection{localEndpoint=[127.0.0.1]:5702, remoteEndpoint=[127.0.0.1]:5701, alive=false}
21:08:31,510  INFO |whenShortTTLAndBigBuffer| - [MembershipManager] main - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] Removing Member [127.0.0.1]:5702 - be2ccebd-6e7a-4b16-8363-00899e7e931d
21:08:31,511  INFO |whenShortTTLAndBigBuffer| - [TransactionManagerService] hz.gracious_fermi.cached.thread-2 - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5702, UUID: be2ccebd-6e7a-4b16-8363-00899e7e931d
21:08:31,511  INFO |whenShortTTLAndBigBuffer| - [ClusterService] main - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] 

Members {size:1, ver:3} [
	Member [127.0.0.1]:5701 - 7ed2c071-d33b-45d0-add3-f748f11ab08f this
]

21:08:31,511  INFO |whenShortTTLAndBigBuffer| - [Node] main - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] Shutting down node engine...
21:08:31,516  INFO |whenShortTTLAndBigBuffer| - [MigrationManager] hz.gracious_fermi.migration - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] Partition balance is ok, no need to repartition.
21:08:31,517  INFO |whenShortTTLAndBigBuffer| - [NodeExtension] main - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] Destroying node NodeExtension.
21:08:31,518  INFO |whenShortTTLAndBigBuffer| - [Node] main - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] Hazelcast Shutdown is completed in 8 ms.
21:08:31,518  INFO |whenShortTTLAndBigBuffer| - [LifecycleService] main - [127.0.0.1]:5702 [dev] [4.2.3-SNAPSHOT] [127.0.0.1]:5702 is SHUTDOWN
21:08:31,518  INFO |whenShortTTLAndBigBuffer| - [LifecycleService] main - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
21:08:31,518  WARN |whenShortTTLAndBigBuffer| - [Node] main - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] Terminating forcefully...
21:08:31,518  INFO |whenShortTTLAndBigBuffer| - [Node] main - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] Shutting down connection manager...
21:08:31,519  INFO |whenShortTTLAndBigBuffer| - [Node] main - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] Shutting down node engine...
21:08:31,522  INFO |whenShortTTLAndBigBuffer| - [NodeExtension] main - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] Destroying node NodeExtension.
21:08:31,524  INFO |whenShortTTLAndBigBuffer| - [Node] main - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] Hazelcast Shutdown is completed in 6 ms.
21:08:31,524  INFO |whenShortTTLAndBigBuffer| - [LifecycleService] main - [127.0.0.1]:5701 [dev] [4.2.3-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
BuildInfo right after whenShortTTLAndBigBuffer(com.hazelcast.ringbuffer.impl.RingbufferAddAllReadManyStressTest): BuildInfo{version='4.2.3-SNAPSHOT', build='20210814', buildNumber=20210814, revision=a516020, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'whenShortTTLAndBigBuffer(com.hazelcast.ringbuffer.impl.RingbufferAddAllReadManyStressTest):'
21:08:15, accumulated pauses: 187 ms, max pause: 117 ms, pauses over 1000 ms: 0
21:08:20, accumulated pauses: 5734 ms, max pause: 5478 ms, pauses over 1000 ms: 1
21:08:30, accumulated pauses: 40 ms, max pause: 32 ms, pauses over 1000 ms: 0
```

--- a/hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferService.java
+++ b/hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferService.java
@@ -215,7 +215,7 @@ public class RingbufferService implements ManagedService, RemoteService, Fragmen
     private Map<ObjectNamespace, RingbufferContainer> getOrCreateRingbufferContainers(int partitionId) {
         final Map<ObjectNamespace, RingbufferContainer> partitionContainer = containers.get(partitionId);
         if (partitionContainer == null) {
-            containers.putIfAbsent(partitionId, new HashMap<>());
+            containers.putIfAbsent(partitionId, new ConcurrentHashMap<>());
         }
         return containers.get(partitionId);
     }

--- a/hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/AbstractRingBufferOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/AbstractRingBufferOperation.java
@@ -27,9 +27,11 @@ import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
 import com.hazelcast.ringbuffer.StaleSequenceException;
 import com.hazelcast.ringbuffer.impl.RingbufferContainer;
 import com.hazelcast.ringbuffer.impl.RingbufferService;
+import com.hazelcast.ringbuffer.impl.RingbufferWaitNotifyKey;
 import com.hazelcast.spi.impl.operationservice.NamedOperation;
 import com.hazelcast.spi.impl.operationservice.Operation;
 import com.hazelcast.spi.impl.operationservice.PartitionAwareOperation;
+import com.hazelcast.spi.impl.operationservice.WaitNotifyKey;
 import com.hazelcast.topic.impl.reliable.ReliableTopicService;
 
 import java.io.IOException;
@@ -88,6 +90,46 @@ public abstract class AbstractRingBufferOperation extends Operation implements N
 
         ringbuffer.cleanup();
         return ringbuffer;
+    }
+
+    /**
+     * Returns an {@link RingbufferContainer} or null if one doesn't exist.
+     * <p>
+     * If it does it also calls the {@link RingbufferContainer#cleanup()} before returning
+     * the container. This will currently remove any expired items.
+     *
+     * @return the ringbuffer container
+     */
+    RingbufferContainer getRingBufferContainerOrNull() {
+        final RingbufferService service = getService();
+        final ObjectNamespace ns = RingbufferService.getRingbufferNamespace(name);
+
+        RingbufferContainer ringbuffer = service.getContainerOrNull(getPartitionId(), ns);
+        if (ringbuffer != null) {
+            ringbuffer.cleanup();
+        }
+
+        return ringbuffer;
+    }
+
+    /**
+     * Returns {@link WaitNotifyKey} of the ringbuffer.
+     *
+     * If the RingbufferContainer exists it reuses it's {@link RingbufferContainer#getRingEmptyWaitNotifyKey()}.
+     * If the RingbufferContainer doesn't exist it creates new RingbufferWaitNotifyKey and doesn't recreate
+     * the ringbuffer container.
+     *
+     * @return WaitNotifyKey of the ringbuffer
+     */
+    WaitNotifyKey getRingbufferWaitNotifyKey() {
+        final RingbufferService service = getService();
+        final ObjectNamespace ns = RingbufferService.getRingbufferNamespace(name);
+        RingbufferContainer ringbuffer = service.getContainerOrNull(getPartitionId(), ns);
+        if (ringbuffer != null) {
+            return ringbuffer.getRingEmptyWaitNotifyKey();
+        } else {
+            return new RingbufferWaitNotifyKey(ns, getPartitionId());
+        }
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java
@@ -17,13 +17,10 @@
 package com.hazelcast.ringbuffer.impl.operations;
 
 import com.hazelcast.core.IFunction;
-import com.hazelcast.internal.services.ObjectNamespace;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.ringbuffer.impl.ReadResultSetImpl;
 import com.hazelcast.ringbuffer.impl.RingbufferContainer;
-import com.hazelcast.ringbuffer.impl.RingbufferService;
-import com.hazelcast.ringbuffer.impl.RingbufferWaitNotifyKey;
 import com.hazelcast.spi.impl.operationservice.BlockingOperation;
 import com.hazelcast.spi.impl.operationservice.ReadonlyOperation;
 import com.hazelcast.spi.impl.operationservice.WaitNotifyKey;
@@ -60,9 +57,7 @@ public class ReadManyOperation<O> extends AbstractRingBufferOperation
 
     @Override
     public boolean shouldWait() {
-        RingbufferService service = getService();
-        ObjectNamespace namespace = RingbufferService.getRingbufferNamespace(name);
-        RingbufferContainer ringbuffer = service.getContainerOrNull(getPartitionId(), namespace);
+        RingbufferContainer ringbuffer = getRingBufferContainerOrNull();
 
         if (resultSet == null) {
             resultSet = new ReadResultSetImpl<>(minSize, maxSize, getNodeEngine().getSerializationService(), filter);
@@ -118,8 +113,7 @@ public class ReadManyOperation<O> extends AbstractRingBufferOperation
 
     @Override
     public WaitNotifyKey getWaitKey() {
-        ObjectNamespace namespace = RingbufferService.getRingbufferNamespace(name);
-        return new RingbufferWaitNotifyKey(namespace, getPartitionId());
+        return getRingbufferWaitNotifyKey();
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadOneOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadOneOperation.java
@@ -43,13 +43,18 @@ public class ReadOneOperation extends AbstractRingBufferOperation implements Blo
 
     @Override
     public void beforeRun() throws Exception {
-        RingbufferContainer ringbuffer = getRingBufferContainer();
-        ringbuffer.checkBlockableReadSequence(sequence);
+        RingbufferContainer ringbuffer = getRingBufferContainerOrNull();
+        if (ringbuffer != null) {
+            ringbuffer.checkBlockableReadSequence(sequence);
+        }
     }
 
     @Override
     public boolean shouldWait() {
-        RingbufferContainer ringbuffer = getRingBufferContainer();
+        RingbufferContainer ringbuffer = getRingBufferContainerOrNull();
+        if (ringbuffer == null) {
+            return true;
+        }
         if (ringbuffer.isTooLargeSequence(sequence) || ringbuffer.isStaleSequence(sequence)) {
             //no need to wait, let the operation continue and fail in beforeRun
             return false;
@@ -66,8 +71,7 @@ public class ReadOneOperation extends AbstractRingBufferOperation implements Blo
 
     @Override
     public WaitNotifyKey getWaitKey() {
-        RingbufferContainer ringbuffer = getRingBufferContainer();
-        return ringbuffer.getRingEmptyWaitNotifyKey();
+        return getRingbufferWaitNotifyKey();
     }
 
     @Override

--- a/hazelcast/src/test/java/com/hazelcast/ringbuffer/impl/RingbufferAddAllReadManyStressTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/ringbuffer/impl/RingbufferAddAllReadManyStressTest.java
@@ -29,11 +29,12 @@ import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.TestThread;
 import com.hazelcast.test.annotation.NightlyTest;
 import org.junit.After;
+import org.junit.Before;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
-import java.util.LinkedList;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Random;
 import java.util.concurrent.ExecutionException;
@@ -58,6 +59,13 @@ public class RingbufferAddAllReadManyStressTest extends HazelcastTestSupport {
     private final AtomicBoolean stop = new AtomicBoolean();
 
     private Ringbuffer<Long> ringbuffer;
+
+    @Before
+    public void setUp() throws Exception {
+        // We sometimes get a GC pause longer than TTL (2s) on IBM JDK
+        // The idea is that we hint GC before the test to clear any garbage from previous tests
+        System.gc();
+    }
 
     @After
     public void tearDown() {
@@ -140,6 +148,7 @@ public class RingbufferAddAllReadManyStressTest extends HazelcastTestSupport {
         private long lastLogMs = 0;
 
         private volatile long produced;
+        private final List<Long> items = new ArrayList<>(MAX_BATCH);
 
         ProduceThread() {
             super("ProduceThread");
@@ -153,7 +162,7 @@ public class RingbufferAddAllReadManyStressTest extends HazelcastTestSupport {
         @Override
         public void doRun() throws Throwable {
             while (!stop.get()) {
-                List<Long> items = makeBatch();
+                makeBatch();
                 addAll(items);
             }
 
@@ -161,9 +170,9 @@ public class RingbufferAddAllReadManyStressTest extends HazelcastTestSupport {
         }
 
         @SuppressWarnings("NonAtomicOperationOnVolatileField")
-        private List<Long> makeBatch() {
+        private void makeBatch() {
+            items.clear();
             int count = max(1, random.nextInt(MAX_BATCH));
-            LinkedList<Long> items = new LinkedList<Long>();
 
             for (int k = 0; k < count; k++) {
                 items.add(produced);
@@ -175,7 +184,6 @@ public class RingbufferAddAllReadManyStressTest extends HazelcastTestSupport {
                     logger.info(getName() + " at " + produced);
                 }
             }
-            return items;
         }
 
         private void addAll(List<Long> items) throws Exception {
@@ -236,6 +244,12 @@ public class RingbufferAddAllReadManyStressTest extends HazelcastTestSupport {
                             throw e;
                         }
                     }
+                }
+
+                if (result.getSequence(0) != seq) {
+                    logger.info("Sequence of the first retrieved result is newer than expected sequence, "
+                            + "this is possible for readManyAsync operation");
+                    seq = result.getSequence(0);
                 }
 
                 for (Long item : result) {

--- a/hazelcast/src/test/java/com/hazelcast/ringbuffer/impl/RingbufferDestroyTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/ringbuffer/impl/RingbufferDestroyTest.java
@@ -1,0 +1,97 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.ringbuffer.impl;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.config.RingbufferConfig;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.internal.services.ObjectNamespace;
+import com.hazelcast.ringbuffer.Ringbuffer;
+import com.hazelcast.test.AssertTask;
+import com.hazelcast.test.HazelcastTestSupport;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Map;
+
+import static com.hazelcast.test.Accessors.getNodeEngineImpl;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+
+public class RingbufferDestroyTest extends HazelcastTestSupport {
+
+    public static final String NAME = "ringbuffer-foo";
+
+    private static HazelcastInstance[] instances;
+    private Ringbuffer<String> ringbuffer;
+
+    @Before
+    public void setUp() throws Exception {
+        Config config = smallInstanceConfig();
+        config.addRingBufferConfig(new RingbufferConfig(NAME).setCapacity(10));
+        instances = createHazelcastInstances(config, 2);
+
+        ringbuffer = instances[0].getRingbuffer(NAME);
+    }
+
+    @Test
+    public void whenDestroyAfterAdd_thenRingbufferRemoved() {
+        ringbuffer.add("1");
+        ringbuffer.destroy();
+
+        assertTrueEventually(new AssertNoRingbufferContainerTask(), 10);
+    }
+
+    @Test
+    public void whenReadOneAfterDestroy_thenMustNotRecreateRingbuffer() {
+        ringbuffer.add("1");
+        ringbuffer.destroy();
+
+        spawn(() -> ringbuffer.readOne(0));
+
+        sleepMillis(100);
+
+        assertTrueEventually(new AssertNoRingbufferContainerTask(), 10);
+    }
+
+    @Test
+    public void whenReadManyAfterDestroy_thenMustNotRecreateRingbuffer() {
+        ringbuffer.add("1");
+        ringbuffer.destroy();
+
+        spawn(() -> ringbuffer.readManyAsync(0, 1, 1, null));
+
+        sleepMillis(100);
+
+        assertTrueEventually(new AssertNoRingbufferContainerTask(), 10);
+    }
+
+    private static class AssertNoRingbufferContainerTask implements AssertTask {
+        @Override
+        public void run() throws Exception {
+            for (HazelcastInstance instance : instances) {
+                final RingbufferService ringbufferService
+                        = getNodeEngineImpl(instance).getService(RingbufferService.SERVICE_NAME);
+
+                final Map<ObjectNamespace, RingbufferContainer> partitionContainers =
+                        ringbufferService.getContainers().get(ringbufferService.getRingbufferPartitionId(NAME));
+                assertNotNull(partitionContainers);
+                assertFalse(partitionContainers.containsKey(RingbufferService.getRingbufferNamespace(NAME)));
+            }
+        }
+    }
+}

--- a/hazelcast/src/test/java/com/hazelcast/ringbuffer/impl/RingbufferTTLTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/ringbuffer/impl/RingbufferTTLTest.java
@@ -19,6 +19,7 @@ package com.hazelcast.ringbuffer.impl;
 import com.hazelcast.config.Config;
 import com.hazelcast.config.RingbufferConfig;
 import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.ringbuffer.ReadResultSet;
 import com.hazelcast.ringbuffer.Ringbuffer;
 import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastParallelClassRunner;
@@ -30,6 +31,7 @@ import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
 import static com.hazelcast.test.Accessors.getNodeEngineImpl;
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNull;
 
@@ -38,7 +40,7 @@ import static org.junit.Assert.assertNull;
 public class RingbufferTTLTest extends HazelcastTestSupport {
 
     private HazelcastInstance hz;
-    private Ringbuffer ringbuffer;
+    private Ringbuffer<String> ringbuffer;
     private RingbufferContainer ringbufferContainer;
     private ArrayRingbuffer arrayRingbuffer;
 
@@ -147,5 +149,21 @@ public class RingbufferTTLTest extends HazelcastTestSupport {
                 }
             }
         }, 5);
+    }
+
+    @Test
+    public void whenTTLEnabled_thenReadManyShouldSkipExpiredItems() throws Exception {
+        setup(new RingbufferConfig("foo").setTimeToLiveSeconds(1).setCapacity(100));
+
+        long head = ringbuffer.headSequence();
+        ringbuffer.add("a");
+
+        ReadResultSet<String> result = ringbuffer.readManyAsync(head, 0, 10, null).toCompletableFuture().get();
+        assertThat(result).containsOnly("a");
+
+        sleepMillis(1100);
+
+        result = ringbuffer.readManyAsync(head, 0, 10, null).toCompletableFuture().get();
+        assertThat(result).isEmpty();
     }
 }

--- a/hazelcast/src/test/java/com/hazelcast/ringbuffer/impl/operations/ReadOneOperationTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/ringbuffer/impl/operations/ReadOneOperationTest.java
@@ -101,7 +101,7 @@ public class ReadOneOperationTest extends HazelcastTestSupport {
 
         ReadOneOperation op = getReadOneOperation(ringbuffer.tailSequence() + 1);
 
-        // since there is an item, we don't need to wait
+        // since there is no item, we should wait
         boolean shouldWait = op.shouldWait();
         assertTrue(shouldWait);
     }

